h2o-automl/src/test/java/ai/h2o/automl/AutoMLModelMetricsTest.java

```diff
@@ -0,0 +1,76 @@
+package ai.h2o.automl;
+
+import hex.Model;
+import hex.splitframe.ShuffleSplitFrame;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import water.Key;
+import water.fvec.Frame;
+
+import static org.junit.Assert.assertEquals;
+
+public class AutoMLModelMetricsTest extends water.TestUtil {
+
+  @BeforeClass public static void setup() { stall_till_cloudsize(1); }
+
+  private Frame getPreparedTitanicFrame(String responseColumnName) {
+    Frame fr = parse_test_file("./smalldata/gbm_test/titanic.csv");
+    fr.remove(new String[]{"name", "ticket", "boat", "body"});
+    asFactor(fr, responseColumnName);
+    return fr;
+  }
+  
+  private Frame[] getSplitsFromTitanicDataset(String responseColumnName) {
+    Frame fr = getPreparedTitanicFrame(responseColumnName);
+
+    double[] ratios = ard(0.7, 0.15, 0.15);
+    Key<Frame>[] keys = aro(Key.<Frame>make(), Key.<Frame>make(), Key.<Frame>make());
+    Frame[] splits = null;
+    splits = ShuffleSplitFrame.shuffleSplitFrame(fr, keys, ratios, 42);
+
+    return splits;
+  }
+  
+  private double getScoreBasedOn(Frame fr, Model model) {
+    model.score(fr);
+    hex.ModelMetricsBinomial mmWithoutTE = hex.ModelMetricsBinomial.getFromDKV(model, fr);
+    return mmWithoutTE.auc();
+  }
+
+  @Test public void scoresOnLeaderBoardFrameAndTheSameFrameAfterwardsShouldBeEqualTest() {
+
+    String responseColumnName = "survived";
+
+    Frame[] splits = getSplitsFromTitanicDataset(responseColumnName);
+
+    AutoMLBuildSpec autoMLBuildSpec = new AutoMLBuildSpec();
+    autoMLBuildSpec.input_spec.training_frame = splits[0]._key;
+    autoMLBuildSpec.input_spec.validation_frame = splits[1]._key;
+    autoMLBuildSpec.input_spec.leaderboard_frame = splits[2]._key;
+    autoMLBuildSpec.build_control.nfolds = 0;
+    autoMLBuildSpec.input_spec.response_column = responseColumnName;
+
+    autoMLBuildSpec.build_control.project_name = "first";
+    autoMLBuildSpec.build_control.stopping_criteria.set_max_models(1);
+    autoMLBuildSpec.build_control.stopping_criteria.set_seed(7890);
+    autoMLBuildSpec.build_control.keep_cross_validation_models = false;
+    autoMLBuildSpec.build_control.keep_cross_validation_predictions = false;
+
+    AutoML aml = AutoML.startAutoML(autoMLBuildSpec);
+    aml.get();
+    Leaderboard leaderboard = aml.leaderboard();
+
+    // It is expected that auc() called on leader will return value estimated on `leaderboard_frame` but it is not the case.
```

Comment:
I don't think expectation is reasonable, the contract of the getLeader method is to return the leader model.

I does seem natural to expect the leader metrics but this should not be handled on the model level.

Reply:
@michalkurka we had this discussion with @deil87 : of course as soon as Leader is just a Model instance, there's no reason to change the metrics, but to avoid the confusion you mention, we could also return a different type of object, probably proxying Model, with:
- `getLeader().auc()`  would return leaderboard metric.
- `getLeader().model.auc()` would return training/validation metric.
the problem is backwards compatibility of course, maybe we should make our users vote for this kind of change...